File: android/app/src/main/java/com/bigdictaphone/app/services/RateLimitService.kt
package com.bigdictaphone.app.services

import android.content.Context
import androidx.datastore.preferences.core.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * Service for managing Gemini API rate limits and request queuing.
 * 
 * Free tier limits for gemini-2.5-flash:
 * - 5 RPM (requests per minute)
 * - 250K TPM (tokens per minute)
 * - 20 RPD (requests per day)
 */
class RateLimitService(private val context: Context) {
    
    private val json = Json { ignoreUnknownKeys = true }
    private val mutex = Mutex()
    
    // Rate limit constants (based on user's dashboard for gemini-2.5-flash)
    companion object {
        const val REQUESTS_PER_MINUTE = 5
        const val REQUESTS_PER_DAY = 20
        const val TOKENS_PER_MINUTE = 250_000
        
        private const val MINUTE_MS = 60_000L
        private const val DAY_MS = 86_400_000L
    }
    
    // In-memory tracking
    private val requestTimestamps = mutableListOf<Long>()
    private var dailyRequestCount = 0
    private var dailyResetTime = 0L
    
    // Request queue
    private val requestQueue = ConcurrentLinkedQueue<QueuedRequest>()
    
    /**
     * Check if we can make an API request right now
     */
    suspend fun canMakeRequest(): RateLimitStatus = mutex.withLock {
        cleanupOldTimestamps()
        updateDailyTracking()
        
        val requestsInLastMinute = requestTimestamps.count { 
            System.currentTimeMillis() - it < MINUTE_MS 
        }
        
        return when {
            dailyRequestCount >= REQUESTS_PER_DAY -> {
                val resetIn = dailyResetTime - System.currentTimeMillis()
                RateLimitStatus.DailyLimitReached(resetIn)
            }
            requestsInLastMinute >= REQUESTS_PER_MINUTE -> {
                val oldestInWindow = requestTimestamps.filter { 
                    System.currentTimeMillis() - it < MINUTE_MS 
                }.minOrNull() ?: System.currentTimeMillis()
                val resetIn = MINUTE_MS - (System.currentTimeMillis() - oldestInWindow)
                RateLimitStatus.MinuteLimitReached(resetIn)
            }
            else -> {
                RateLimitStatus.Available(
                    remainingPerMinute = REQUESTS_PER_MINUTE - requestsInLastMinute,
                    remainingPerDay = REQUESTS_PER_DAY - dailyRequestCount
                )
            }
        }
    }
    
    /**
     * Record that a request was made
     */
    suspend fun recordRequest() = mutex.withLock {
        val now = System.currentTimeMillis()
        requestTimestamps.add(now)
        dailyRequestCount++
        
        // Keep only timestamps from the last minute for efficiency
        cleanupOldTimestamps()
    }
    
    /**
     * Get wait time until next available slot (in milliseconds)
     */
    suspend fun getWaitTime(): Long {
        return when (val status = canMakeRequest()) {
            is RateLimitStatus.Available -> 0L
            is RateLimitStatus.MinuteLimitReached -> status.resetInMs
            is RateLimitStatus.DailyLimitReached -> status.resetInMs
        }
    }
    
    /**
     * Add a request to the queue
     */
    fun queueRequest(request: QueuedRequest) {
        requestQueue.add(request)
    }
    
    /**
     * Get next queued request (if any)
     */
    fun getNextQueuedRequest(): QueuedRequest? {
        return requestQueue.poll()
    }
    
    /**
     * Get queue size
     */
    fun getQueueSize(): Int = requestQueue.size
    
    /**
     * Check if queue has items
     */
    fun hasQueuedRequests(): Boolean = requestQueue.isNotEmpty()
    
    /**
     * Clean up timestamps older than 1 minute
     */
    private fun cleanupOldTimestamps() {
        val cutoff = System.currentTimeMillis() - MINUTE_MS
        requestTimestamps.removeAll { it < cutoff }
    }
    
    /**
     * Update daily tracking (reset if new day)
     */
    private fun updateDailyTracking() {
        val now = System.currentTimeMillis()
        if (dailyResetTime == 0L || now >= dailyResetTime) {
            // Reset daily counter at midnight
            val calendar = Calendar.getInstance().apply {
                add(Calendar.DAY_OF_YEAR, 1)
                set(Calendar.HOUR_OF_DAY, 0)
                set(Calendar.MINUTE, 0)
                set(Calendar.SECOND, 0)
                set(Calendar.MILLISECOND, 0)
            }
            dailyResetTime = calendar.timeInMillis
            dailyRequestCount = 0
        }
    }
    
    /**
     * Get current rate limit status as a displayable string
     */
    suspend fun getStatusString(): String {
        return when (val status = canMakeRequest()) {
            is RateLimitStatus.Available -> 
                "✅ ${status.remainingPerMinute}/min • ${status.remainingPerDay}/day"
            is RateLimitStatus.MinuteLimitReached -> 
                "⏳ Retry in ${status.resetInMs / 1000}s"
            is RateLimitStatus.DailyLimitReached -> 
                "🚫 Daily limit reached"
        }
    }
}

/**
 * Rate limit status
 */
sealed class RateLimitStatus {
    data class Available(
        val remainingPerMinute: Int,
        val remainingPerDay: Int
    ) : RateLimitStatus()
    
    data class MinuteLimitReached(val resetInMs: Long) : RateLimitStatus()
    data class DailyLimitReached(val resetInMs: Long) : RateLimitStatus()
}

/**
 * Request in the queue
 */
@Serializable
data class QueuedRequest(
    val id: String = UUID.randomUUID().toString(),
    val recordingId: String,
    val timestamp: Long = System.currentTimeMillis()
)

File: Big-DicTaphone/Services/RateLimitService.swift
import Foundation

/// Service for managing Gemini API rate limits and request queuing.
/// Free tier limits for gemini-2.5-flash:
/// - 5 RPM (requests per minute)
/// - 20 RPD (requests per day)
actor RateLimitService {
    static let shared = RateLimitService()
    
    // Rate limit constants
    private let requestsPerMinute = 5
    private let requestsPerDay = 20
    private let minuteInterval: TimeInterval = 60
    
    // Storage keys
    private let timestampsKey = "api_request_timestamps"
    private let dailyCountKey = "api_daily_count"
    private let dailyResetKey = "api_daily_reset_time"
    
    private let defaults = UserDefaults.standard
    
    private init() {}
    
    enum RateLimitStatus {
        case available
        case minuteLimitReached(resetIn: TimeInterval)
        case dailyLimitReached
    }
    
    /// Check if we can make a request
    func canMakeRequest() -> RateLimitStatus {
        cleanupOldTimestamps()
        updateDailyTracking()
        
        let dailyCount = defaults.integer(forKey: dailyCountKey)
        if dailyCount >= requestsPerDay {
            return .dailyLimitReached
        }
        
        let timestamps = (defaults.array(forKey: timestampsKey) as? [TimeInterval]) ?? []
        if timestamps.count >= requestsPerMinute {
            if let oldest = timestamps.first {
                let now = Date().timeIntervalSince1970
                let resetIn = minuteInterval - (now - oldest)
                return .minuteLimitReached(resetIn: max(0, resetIn))
            }
        }
        
        return .available
    }
    
    /// Record a request
    func recordRequest() {
        let now = Date().timeIntervalSince1970
        
        // Update minute tracking
        var timestamps = (defaults.array(forKey: timestampsKey) as? [TimeInterval]) ?? []
        timestamps.append(now)
        defaults.set(timestamps, forKey: timestampsKey)
        
        // Update daily tracking
        let dailyCount = defaults.integer(forKey: dailyCountKey)
        defaults.set(dailyCount + 1, forKey: dailyCountKey)
        
        cleanupOldTimestamps()
    }
    
    /// Clean up timestamps older than 1 minute
    private func cleanupOldTimestamps() {
        let now = Date().timeIntervalSince1970
        let cutoff = now - minuteInterval
        
        var timestamps = (defaults.array(forKey: timestampsKey) as? [TimeInterval]) ?? []
        timestamps = timestamps.filter { $0 >= cutoff }
        defaults.set(timestamps, forKey: timestampsKey)
    }
    
    /// Reset daily counter if a new day has started
    private func updateDailyTracking() {
        let now = Date().timeIntervalSince1970
        let resetTime = defaults.double(forKey: dailyResetKey)
        
        if resetTime == 0 || now >= resetTime {
            // Calculate next midnight
            var calendar = Calendar.current
            if let tomorrow = calendar.date(byAdding: .day, value: 1, to: Date()),
               let midnight = calendar.date(bySettingHour: 0, minute: 0, second: 0, of: tomorrow) {
                
                defaults.set(midnight.timeIntervalSince1970, forKey: dailyResetKey)
                defaults.set(0, forKey: dailyCountKey)
            }
        }
    }
    
    /// Get status string for UI
    func getStatusString() -> String {
        let status = canMakeRequest()
        let timestamps = (defaults.array(forKey: timestampsKey) as? [TimeInterval]) ?? []
        let dailyCount = defaults.integer(forKey: dailyCountKey)
        
        switch status {
        case .available:
            return "✅ \(requestsPerMinute - timestamps.count)/min • \(requestsPerDay - dailyCount)/day"
        case .minuteLimitReached(let resetIn):
            return "⏳ Retry in \(Int(resetIn))s"
        case .dailyLimitReached:
            return "🚫 Daily limit reached"
        }
    }
}
